Data port (or update smartport) Power line On/Off

dgatf opened the pull request:
Hi,

Trying to make my first PR to this awesome project... so it might not be perfect

This is related to:

[https://github.com/opentx/opentx/issues/4838](url)
[https://github.com/opentx/opentx/issues/6614](url)

Adding switch control to the power line to the port adds two new functions to the port:

- Power any device attached to the port (5V) (4838)
- Allow to connect a bluetooth module like HC-05 or HC-06 to passthrough the telemetry (6614)

A new checkbox is added to the _hardware menu_, _SP Power_, to control the data line power on/off. AFAIK this interface present in X7, X9Lite, XLite/S andX10. I've implemented for X7, X9Lite and XLite/S (not for X10 yet). No auto feature as pointed in 6614

Tested telemetry passthrough with X7, HC-05 and 4049 signal inverter. On android side using iNav. All working good 

Also tested smartport firmware update (x8r) with and without power on the port before updating with no issue

I've the following remarks:

- Changing the state of the line is done in _radio_hardware.cpp_, not sure if that is the best place
- There is an issue when scrolling over the check box as it needs two clicks on the rotary instead of one

Let me know any comment, hopefully this is not too bad

Cheers,
Daniel

dgatf committed:
SmartPort Power

dgatf committed:
SmartPort Power

dgatf committed:
Add data power control

dgatf committed:
Add power control to data port

dgatf commented on radio/src/gui/common/stdlcd/radio_hardware.cpp:
```diff
@@ -631,6 +634,15 @@ void menuRadioHardware(event_t event)
+        if (g_eeGeneral.smartportUpdatePower == 0)
+          sportUpdatePowerOn();
+        else
+          sportUpdatePowerOff();
```
Is here the best place to update the power line state?

dgatf commented on radio/src/datastructs.h:
```diff
@@ -917,28 +918,28 @@ static inline void check_struct()
   CHKSIZE(TrainerData, 16);
 
 #if defined(PCBXLITES)
-  CHKSIZE(RadioData, 860);
```
Increased size by 1. Is all that is needed?

dgatf commented on radio/src/gui/common/stdlcd/radio_hardware.cpp:
```diff
@@ -631,6 +634,15 @@ void menuRadioHardware(event_t event)
         break;
+#if defined(PCBX7) || defined(PCBXLITE) || defined(PCBX9LITE) || defined(PCBXLITES)
+       case ITEM_RADIO_HARDWARE_SMARTPORT_POWER:
+        g_eeGeneral.smartportUpdatePower = 1 - editCheckBox(1 - g_eeGeneral.smartportUpdatePower, HW_SETTINGS_COLUMN2, y, "SP power", attr, event);
```
There is an issue when scrolling over the check box as it needs two clicks on the rotary instead of one

dgatf commented:
Faling checks at _Conversion From22_
Could be related to the increased size of the RadioData (EEPROM)?
No idea how to solve this

kilrah commented:
The data structure cannot change at any time other than from a major version to the next (e.g. 2.3 -> 2.4). So unless there is a spare bit that can be used, and with which there is no more interesting thing to do then a feature that needs a setting cannot be introduced just at any time.

dgatf committed:
Add power control to data port

dgatf commented:
Ok. Now using spare3 from RadioData

kilrah commented:
> No auto feature as pointed in 6614

Auto wasn't a feature, it was the label to use instead of OFF. Point is, when you run a firmware update operation on the port the power will be switched on/off as necessary.
So instead of a checkbox I'd see text choices ON/AUTO (translatable of course).

Becasue of the firmware update taking control of sport power just looking at the setting at power up isn't good, as it is now if your setting is on power will be turned on at power up, but if you then do a device firmware update it will get turned off.

3djc commented:
Your menu navigation is broken because you add a line, yet you do not tell the menu handler (arround line 340), but be aware that you CANNOT use "#if defined()" in the menu list itself, you need to do it outside of it

dgatf committed:
Add power control to data port

dgatf committed:
Add power control to data port

dev-fred commented:
I have a X10 express and I am very interested by this PR when you will add X10.
I use also a telemetry passthrough = HC-06 + sport inverter connected to R9M sport connector and on android side using CrazyDude application to get inav telemetry. 
The best would be to use the sport connector on the radio today specialized for firmware flashing .

dgatf committed:
Add power control to data port

dgatf commented:
Thanks kilrah, 3djc. I've made the following changes:

- Changed to translatable editChoice 'on/auto' 
- Added function sportUpdatePowerInit to the Sport update driver to update the state of the power line based on the config
- Added macro IS_SPORT_UPDATE_POWER_ON to the Sport update driver to check the state of the power line
- Power line state is updated at opentxInit and radio hardware menu
- Power line is left in the same state after frsky firmware update and multi firmware update 
- Fixed broken menu
- Support for all taranis/horus boards with this port

3djc commented on radio/src/gui/480x272/radio_hardware.cpp:
```diff
@@ -365,6 +375,16 @@ bool menuRadioHardware(event_t event)
+#if HAS_SPORT_UPDATE_CONNECTOR()
+      case ITEM_RADIO_HARDWARE_SPORT_UPDATE_POWER:
+        lcdDrawText(MENUS_MARGIN_LEFT, y, STR_SPORT_UPDATE_POWER_MODE);
+        g_eeGeneral.sportUpdatePower = editChoice(HW_SETTINGS_COLUMN+50, y, STR_SPORT_UPDATE_POWER_MODES, g_eeGeneral.sportUpdatePower, 0, 1, attr, event);
```
Spaces ‘ + 50’

3djc commented on radio/src/io/frsky_firmware_update.cpp:
```diff
@@ -752,8 +757,9 @@ const char * FrskyChipFirmwareUpdate::flashFirmware(const char * filename, bool
+  uint8_t spuPwr = IS_SPORT_UPDATE_POWER_ON();
   SPORT_UPDATE_POWER_OFF();
-
+  
```
Cosmetics

dgatf replied:
not sure...name for spuPwr or else?

3djc replied:
the space character

3djc commented on radio/src/opentx.cpp:
```diff
@@ -1966,6 +1966,10 @@ void opentxInit()
   btInit();
 #endif
 
+#if defined(PCBTARANIS) || defined(PCBHORUS)
```
#if defined(SPORT_UPDATE_PWR_GPIO) ?

dgatf replied:
Yes, that's better. More generic

dgatf committed:
Change sport power init condition

dgatf committed:
Cosmetics

dgatf committed:
Cosmetics

dgatf commented:
Travis seems to be timed out and produced an errored check
Do I need to do another commit or something?

3djc commented:
I have manually restarted those

dgatf commented:
> I have a X10 express and I am very interested by this PR when you will add X10.
> I use also a telemetry passthrough = HC-06 + sport inverter connected to R9M sport connector and on android side using CrazyDude application to get inav telemetry.
> The best would be to use the sport connector on the radio today specialized for firmware flashing .

I've added support for any taranis/horus with this port. Yes, this is for the connector used for firmware updates (3 pins). This port already have smartport telemetry. It was only needed to add power control to feed the bluetooth or any other device

Go to _hardware menu -> SP power -> ON_
AUTO is for the current mode

kilrah commented:
Actually on X10 Express it's a whole other problem since the internal module doesn't use the SPORT line anymore and the protocol has changed too.

dev-fred commented:
As I understand it, 
there's no way to retrieve the telemetry from the ISRM. 
- either via the Sport connector on the radio
- or via the internal bluetooth (which doesn't seem to be working, by the way).

On the other hand, the Sport pin of the X10E connector is wired to the Sport pin of the external module whether it's an R9M or an XJT,
The only pin switchable by software is the power supply on the connector of the X10E, 
the possibility to control this power supply is a must, 
it's easier to have a bluetooth module HC-05/06 attached to the radio than to the external module which can often be unplugged and what about the multimodule which doesn't have a Sport connector.
I hope that make sens for you

kilrah commented:
Well that's the whole point of this pull request, so not sure why you seem to want to convince anyone?

3djc commented on radio/src/datastructs.h:
```diff
@@ -941,4 +941,4 @@ static inline void check_struct()
 
 #undef CHKSIZE
 }
-#endif /* BACKUP */
```
Cosmetic (space)

dev-fred commented:
Sorry Kilrah, I just was not sure if the X10E will benefit of this PR :"I've implemented for X7, X9Lite and XLite/S (not for X10 yet). "

kilrah commented:
He had not done YET but added it before your comment.

> I've added support for any taranis/horus with this port.

dev-fred commented:
A big thanks to dgatf and the devs, this PR is a very good workaround of the X10E bugged internal BT.

dgatf committed:
Cosmetic

dgatf committed:
Update readme

bsongis commented:
I promised it to @robthomson long time ago, thanks for this PR!

bsongis commented:
I have done some cosmetics, reverted the README.md and changed a definition in another PR, please tell-me if I did wrong. Thanks for your help on this one!

dgatf commented:
Ok. Thanks

dgatf commented:
What is the name of the new PR? My fork has dissapeared. Why? What I did wrong? There was also another branch with spanish language being updated and is algo gone. Only smartport_power branch should been deleted. Maybe I should only do one branch per fork. I was doing forks in two levels (upper is sync with upstream and lowers are the PRs) to keep updated the code faster. Sorry I'm newbie on this
For info, I still have a local copy if needed

kilrah commented:
Looks like you deleted your fork, even if there was specific safeguards before doing so...
There is no reason to make multiple forks.

dgatf commented:
I dont't see the fork since you close this PR. I don't think I've deleted it.. Do you still have the PR somewhere?

3djc commented:
Bsongis took your PR and merged it but had to create a local branch because your last commit would have broken our readme if we did merge your branch

dgatf commented:
Ok, understood. My mistake. Still learning the process. I guess I should have closed the PR before if I wanted to change the readme. Thanks

kilrah commented:
If you want to change the README the commit doing that should be on the repository's default branch. Then you switch to another branch to commit the changes that should be part of the PR.

dgatf commented:
Ok. Thanks
